### PR TITLE
Implement support for container and CAN-FD frames

### PR DESCRIPTION
**Pull request by AlexMicWalz**
- implement support for container-frames,
- implement support for CAN-FD frames,
- adapt output message for loading files.

Alexander Walz &lt;alexander.walz@mercedes-benz.com&gt;, on behalf of [MBition GmbH](https://mbition.io/).
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

### CODE REVIEW

**Comment by AlexMicWalz** [comment]
Hello, meanwhile I've also signed the Eclipse Contributor Agreement, hope this helps.
Kind regards, Alex

**Comment by lukasmittag** [comment]
to get the CI clear you maybe need to rebase and force push or add another commit

**Comment by andlaus** [comment]
is there a way to fix these pesky formatting errors automatically? (something like `ruff --fix`, but `flake8` does not seem to have such a feature...)

**Comment by SebastianSchildt** [comment]
> is there a way to fix these pesky formatting errors automatically? (something like `ruff --fix`, but `flake8` does not seem to have such a feature...)

yes there is!

You do `pip install pre-commit`  in the working directory, so then when you commit, it will check, and fail if something is wrong. You can also run pre-commit in the terminal. it will then modify the files so they match style. Then you jsut need to git add them to accept taht, and commit will work, and also the CI check after push.

**Comment by andlaus** [comment]
I did this for the HEAD of this PR, but `flake8` still complains:
```
kuksa.val.feeders/dbc2val|7a0f8a2... > pre-commit && flake8 .
Trim Trailing Whitespace.............................(no files to check)Skipped
Fix End of Files.....................................(no files to check)Skipped
Check Yaml...........................................(no files to check)Skipped
Check for added large files..........................(no files to check)Skipped
flake8...............................................(no files to check)Skipped
./dbcfeederlib/canreader.py:35:110: E231 missing whitespace after ':'
./dbcfeederlib/canreader.py:35:115: E252 missing whitespace around parameter equals
./dbcfeederlib/canreader.py:35:116: E252 missing whitespace around parameter equals
./dbcfeederlib/canreader.py:35:121: E501 line too long (122 > 120 characters)
./dbcfeederlib/canreader.py:47:121: E501 line too long (132 > 120 characters)
./dbcfeederlib/canreader.py:90:13: F841 local variable 'e' is assigned to but never used
./dbcfeederlib/canreader.py:103:41: E231 missing whitespace after ','
./dbcfeederlib/canreader.py:108:5: E303 too many blank lines (2)
./dbcfeederlib/dbcparser.py:121:1: W293 blank line contains whitespace
./dbcfeederlib/dbcreader.py:35:121: E501 line too long (131 > 120 characters)
```

the various sub-commands for `pre-commit` (e.g., `pre-commit run`) also do not seem to make a difference. what am I missing?

**Comment by andlaus** [comment]
(note: runninng `black` makes `flake8` happy, bit it leads to a *ton* of irrelevant formatting changes. I suppose you're not fond of these?)

**Comment by SebastianSchildt** [comment]
Ahh, I think the issue is, you committed already, hence when you run pe-commit it always determines `(no files to check)Skipped`.

I guess @erikbosch  knows more about the pre-commit hook, but as an ugly way try this
 - Touch you files again, i.e. add a space somehwere
 - git add them
 - run pre-commit
 - it should have modified your files, check whether you like it
 - git add again
 - commit

**Comment by andlaus** [comment]
okay, tried that: unfortunately, it didn't change anything. That said, what formatter tool is `pre-commit` supposed to call in order to fix the coding style? (it could be any of `yapf`, `black`, `autopep8`, ...)

**Comment by SebastianSchildt** [comment]
For the config see https://github.com/eclipse/kuksa.val.feeders/blob/main/.pre-commit-config.yaml and https://github.com/eclipse/kuksa.val.feeders/blob/main/.flake8 

I am currently not sure whether the pre-commit flake part will also modify the source.  The whitespace checker definitely does. If in doubt I think @erikbosch  might support by adding a commit on top of your work

Apart from "style and beauty", functionality wise @erikbosch  might have a say and @sophokles73 as your are an active user of this provider, you might want to check if this still work for your use cases

File: dbc2val/dbcfeederlib/canreader.py [hunk 1]
```diff
@@ -83,44 +83,58 @@ def stop(self):
...
+                log.debug("Decoded message: %s", str(decode))
+            rx_time = time.time()
+
+            if isinstance(decode, dict):
+                # handle normal frame
+                self._handle_decoded_frame(message_def, decode, rx_time)
+            else:
+                # handle container frame
+                for tmp in decode:
+                    if isinstance(tmp[1],bytes):
```
Comment (sophokles73) [inline dbc2val/dbcfeederlib/canreader.py:103]: FMPOV a comment regarding the underlying condition that you are checking here would make this easier to understand
Also, there is a space character missing after the comma ...

File: dbc2val/dbcfeeder.py [hunk 2]
```diff
@@ -451,6 +452,11 @@ def _get_command_line_args_parser() -> argparse.ArgumentParser:
         action="store_true",
         help="Use SocketCAN (overriding any use of --dumpfile)",
     )
+    parser.add_argument(
```
Comment (sophokles73) [inline dbc2val/dbcfeeder.py:455]: this new parameter should probably also be documented in the README

**Comment by sophokles73** [comment]
@AlexMicWalz this looks good to me :-) It would be great if you could also add a unit test checking the FD frame parsing/handling ...

File: dbc2val/dbcfeeder.py (hunk 2 above)
Comment (erikbosch) [reply dbc2val/dbcfeeder.py:455]: Yes, and we should better discuss if we think this one need to be controllable by environment variable and config file as well, similar to how canport is managed.

```
if args.canport:
        canport = args.canport
    elif os.environ.get("CAN_PORT"):
        canport = os.environ.get("CAN_PORT")
    else:
        canport = config.get(CONFIG_SECTION_CAN, CONFIG_OPTION_PORT, fallback=None)

```

**Comment by erikbosch** [comment]
I did some smoketest using the same logfile as we use for release testing and did not observe anything odd. I think we shall add support for defining this in at least config file as we have said in VSS_project that all settings should be available in config file, but command line arg shall have precedence.

**Comment by andlaus** [comment]
> I did some smoketest using the same logfile as we use for release testing and did not observe anything odd

testing for container frames is quite a bit of effort because DBC does not support this feature (you need to use an ARXML file). We'll  look into providing a unit test...

**Comment by andlaus** [comment]
> The current status of the pre-commit checks is that we have not prescribed/proposed any tool to fix the findings. Any suggestion on a good candidate is welcome, as we then could discuss if we should make that one the "recommended" formatting tool.

the approach used by [odxtools](https://github.com/mercedes-benz/odxtools) is to run YAPF as part of the CI and bail out if any discrepancy is detected. To fix such issues a small [shell script](https://github.com/mercedes-benz/odxtools/blob/main/reformat-source.sh) is provided...

**Comment by sophokles73** [comment]
> testing for container frames is quite a bit of effort because DBC does not support this feature (you need to use an ARXML file). We'll look into providing a unit test...

Maybe you can use a similar approach as in https://github.com/eclipse/kuksa.val.feeders/blob/91c3091a8918eed7d79500892ba8d894930b7a67/dbc2val/test/test_readers/test_readers.py#L72

without the need for a DBC file.

**Comment by AlexMicWalz** [comment]
I have fixed the findings of flake8.
Unit-test is in the making.

**Comment by erikbosch** [comment]
We are migrating CAN Provider (dbcfeeder) to a new [repo](https://github.com/eclipse-kuksa/kuksa-can-provider). Discussion may continue in this PR, but there is no point in updating this PR, if so rather create a new PR in the new repository.

**Comment by SebastianSchildt** [comment]
@erikbosch what is the state of this, would be a pity to let it rot. Could you or maybe the @andlaus make an effort to create move the PR to the new location? I feel it was already (almost?) good enough content wise

**Comment by andlaus** [comment]
as far as I remember the main showstopper here was the fact that there was no unit test. I promised one, but when I did so, I assumed that there was a unit test for the  (non-container aware) DBC feeder. seems like this assumption was not correct and after bouncing my head against making one from scratch I gave up. (the problem is how to check that a given CAN signal gets exposed on the Kuksa.val site without having to set up a metric ton of servers, databases, etc.)

I came to the conclusion that it is best to tackle the implementation of end to end unit tests for the feeders in another PR sometime in the future. (I'll go over this PR to make it merge-ready tomorrow or early next week.)

**Comment by andlaus** [comment]
transferred the PR to the new repository: https://github.com/eclipse-kuksa/kuksa-can-provider/pull/6

this PR can be closed AFAICS. (I don't have permission to do so...)

**Pull request closed without merging**